### PR TITLE
Feature/neoepitope

### PR DESCRIPTION
**Pull request by Zethson**
## PR checklist
 - [x] This comment contains a description of changes (with reason)
 - [x] Ensure the test suite passes (`nextflow run . -profile test,docker`).
 - [x] Make sure your code lints (`nf-core lint .`).
 - [x] Documentation in `docs` is updated
 - [ ] `CHANGELOG.md` is updated

This PR is too big, since it also incorporates the changes in the other PR. We should merge the other PR first and then examine this one.
Let's do this together, soon.

This PR enables automated neoepitope resolvement by comparing the found epitopes from the mass spec data to the possible neoepitopes generated from the vcf file. Unique epitopes found in both data sources are labeled neoepitopes and written into a csv file (and alternatively a raw or json file).

Currently the vcf neoepitope script has to use bimas, since MHCFlurry is still not implemented in Fred2.

### CODE REVIEW

**Comment by Zethson** [comment]
Example output:
```
[warm up] executor > local
[78/e40176] Submitted process > output_documentation
[39/498f80] Submitted process > predict_possible_neoepitopes (1)
[14/e007c4] Submitted process > generate_proteins_from_vcf (1)
[1f/8fe1b2] Submitted process > generate_decoy_database (1)
[e0/b904e3] Submitted process > db_search_comet (2)
[19/c23dab] Submitted process > db_search_comet (1)
[2f/df64eb] Submitted process > db_search_comet (3)
[ab/e07b78] Submitted process > db_search_comet (4)
[6a/ca4694] Submitted process > index_peptides (1)
[df/37d524] Submitted process > index_peptides (2)
[8a/00f163] Submitted process > index_peptides (3)
[28/68feec] Submitted process > index_peptides (4)
[c8/ac53d5] Submitted process > calculate_fdr_for_idalignment (1)
[53/677b33] Submitted process > calculate_fdr_for_idalignment (2)
[c7/285d25] Submitted process > calculate_fdr_for_idalignment (3)
[ba/494070] Submitted process > calculate_fdr_for_idalignment (4)
[e5/b6a3d5] Submitted process > filter_fdr_for_idalignment (1)
[d6/5662a7] Submitted process > filter_fdr_for_idalignment (2)
[69/ac1130] Submitted process > filter_fdr_for_idalignment (3)
[2e/2b1e4e] Submitted process > filter_fdr_for_idalignment (4)
[cd/b314dc] Submitted process > align_ids
[5b/1c54f8] Submitted process > align_idxml_files (2)
[7a/003c43] Submitted process > align_idxml_files (1)
[0b/0ad95e] Submitted process > align_mzml_files (1)
[6d/2180ed] Submitted process > align_mzml_files (3)
[42/698321] Submitted process > align_mzml_files (2)
[2e/275743] Submitted process > align_idxml_files (4)
[05/080af5] Submitted process > align_idxml_files (3)
[ea/31a6aa] Submitted process > align_mzml_files (4)
[b4/9e45d5] Submitted process > merge_aligned_idxml_files
[86/e2d5f6] Submitted process > extract_psm_features_for_percolator
[2b/dc9b04] Submitted process > run_percolator
[fc/bc1fd7] Submitted process > filter_by_q_value
[fe/2cde67] Submitted process > quantify_identifications_targeted (2)
[b3/0def0f] Submitted process > quantify_identifications_targeted (1)
[5f/d5588c] Submitted process > quantify_identifications_targeted (3)
[b7/eb9c9a] Submitted process > quantify_identifications_targeted (4)
[4b/bae190] Submitted process > link_extracted_features
[5f/388be4] Submitted process > resolve_conflicts
[e1/1e501a] Submitted process > export_text
[e9/67cfc3] Submitted process > export_mztab
[de/60bbea] Submitted process > predict_peptides (1)
2019-05-05 12:44:53,357 - MHCFlurry Predict mztab - WARNING - Allele: C*02:02 is not supported by MHCFlurry!
[35/0dbf27] Submitted process > Resolve_found_neoepitopes (1)
2019-05-05 12:45:02,662 - Found Neoepitopes - WARNING - Allele B*35:03 was specified, but not found in the possible neoepitopes predicted from the vcf file. No neoepitopes will be resolved for allele: B*35:03
[nf-core/mhcquant] Pipeline Complete
WARN: To render the execution DAG in the required format it is required to install Graphviz -- See http://www.graphviz.org for more info.
```

**Comment by Zethson** [comment]
image

found_neoepitopes.csv

File: docs/usage.md [hunk 1]
```diff
@@ -218,14 +218,19 @@ Size of Spectrum batch for Comet processing (Decrease/Increase depending on Memo
...
 ### `--run_prediction`
-Set to 'True' or 'False' depending on whether binding predictions using the tool mhcflurry should be run. (Check whether your alleles are supported by mhcflurry)
+Set to 'True' or 'False' depending on whether binding predictions using the tool mhcflurry should be run. (Check whether your alleles are [supported by mhcflurry](supported_alleles.md))
 
 ### `--refine_fdr_on_predicted_subset`
-Set to 'True' or 'False' depending on whether binding predictions using the tool mhcflurry should be used to subset all PSMs not passing the q-value threshold. If specified the FDR will be refined using Percolator on the subset of predicted binders among all PSMs resulting in an increased identification rate. (Check whether your alleles are supported by mhcflurry)
+Set to 'True' or 'False' depending on whether binding predictions using the tool mhcflurry should be used to subset all PSMs not passing the q-value threshold. If specified the FDR will be refined using Percolator on the subset of predicted binders among all PSMs resulting in an increased identification rate. (Check whether your alleles are [supported by mhcflurry](supported_alleles.md))
```
Comment (Leon-Bichmann) [inline docs/usage.md]: nice

File: docs/usage.md [hunk 2]
```diff
@@ -218,14 +218,19 @@ Size of Spectrum batch for Comet processing (Decrease/Increase depending on Memo
...
 
 ### `--affinity_threshold_subset`
 Affinity threshold (nM) used to define binders for PSM subset selection in the fdr refinement procedure (eg. 500)
 
+### '--resolve_neoepitopes'
```
Comment (Leon-Bichmann) [inline docs/usage.md]: Looks good, lets discuss this option tomorrow

File: nextflow.config [hunk 3]
```diff
@@ -51,52 +48,53 @@ params {
...
+} catch (Exception e) {
+  System.err.println("WARNING: Could not load nf-core/config profiles: ${params.custom_config_base}/nfcore_custom.config")
+}
 
 profiles {
+  awsbatch { includeConfig 'conf/awsbatch.config' }
+  debug { process.beforeScript = 'echo $HOSTNAME' }
   docker {
     docker.enabled = true
-    process.container = params.container
```
Comment (Leon-Bichmann) [inline nextflow.config]: hmm.. , did this cause any problems or why remove it?

File: nextflow.config [hunk 4]
```diff
@@ -51,52 +48,53 @@ params {
...
     docker.temp = 'auto'
     env {
       JOBLIB_TEMP_FOLDER="/tmp"
     }
   }
-  singularity {
```
Comment (Leon-Bichmann) [inline nextflow.config]: same here?

File: main.nf [hunk 5]
```diff
@@ -255,10 +261,11 @@ if( params.include_proteins_from_vcf){
     Channel
         .fromPath( params.vcf )
         .ifEmpty { exit 1, "params.vcf was empty - no input file supplied" }
-        .set { input_vcf}
+        .into { input_vcf; input_vcf_neoepitope}
```
Comment (Leon-Bichmann) [inline main.nf]: well done

File: bin/vcf_neoepitope_predictor.py [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+                    proteins.append(
+                        Protein(protein_seq, gene_id=l.strip(), transcript_id=ensembl_ids[EAdapterFields.TRANSID]))
+        epitopes = []
+        for length in range(args.peptide_min_length, args.peptide_max_length):
+            epitopes.extend(generate_peptides_from_proteins(proteins, length))
+
+    # read in allele list
+    alleles = read_lines(args.alleles, in_type=Allele)
+
+    result = EpitopePredictorFactory(args.method).predict(epitopes, alleles=alleles)
```
Comment (Leon-Bichmann) [inline bin/vcf_neoepitope_predictor.py]: we might not need to predict the neoepitopes, as there exist some that are not predicted binders.

File: bin/found_neoepitopes.py [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+
+def parse_vcf_neoepitopes(neoepitope_file, alleles):
+    """
+    parses the output of the VCF neoepitope predictor script
+
+    :param neoepitope_file: output file of VCF neoepitope script
+    :param alleles: all alleles which were used for the VCF neoepitope script
+    :return: dictionary of alleles to peptide sequences
+    """
+    HLA_allele_to_peptides = defaultdict(list)
```
Comment (Leon-Bichmann) [inline bin/found_neoepitopes.py]: Nice work, but I wonder whether it might be better to do that independent of the allele

File: conf/test.config [hunk 8]
```diff
@@ -4,7 +4,10 @@
  * -------------------------------------------------
  * Defines bundled input files and everything required
  * to run a fast and simple test. Use as follows:
- *   nextflow run nf-core/methylseq -profile test
+ *   nextflow run nf-core/mhcquant -profile test
```
Comment (Leon-Bichmann) [inline conf/test.config]: extra points, you found the easter egg ! ^^

File: nextflow.config (hunk 3 above)
Comment (Zethson) [reply nextflow.config]: Yes.
This is discouraged now and linting was complaining.
See https://github.com/nf-core/mhcquant/pull/58

**Comment by Zethson** [comment]
@Leon-Bichmann This PR also fixes https://github.com/nf-core/mhcquant/issues/42 .

But let's discuss that tomorrow ;)

**Comment by Zethson** [comment]
Neoepitopes are now automatically resolved when a vcf file was provided.

Bindings are no longer predicted on neoepitopes.
As soon as FRED2 got another release I will readd the binding prediction for neoepitopes.

Also, the code got pretty ugly now. I will think of ways to clean it up, but it's likely easiest to do so when I get my hands back on the binding prediction.

**Comment by Zethson** [comment]
Update: I've enabled neoepitope binding affinity prediction. 
If a vcf file is supplied and run_bindings is enabled, then the neoepitopes will also be subject to affinity binding predictions.

I suggest that we discuss this in person tomorrow. I can clean up the code afterwards.

**Comment by Leon-Bichmann** [comment]
sounds good!

File: main.nf [hunk 9]
```diff
@@ -1024,6 +1024,76 @@ process predict_peptides {
...
+
+    input:
+     file mztab from features_mztab_neoepitopes
+     file neoepitopes from possible_neoepitopes
+
+    output:
+     file "found_neoepitopes.csv" into found_neoepitopes
+    
+    when:
+     params.vcf.length() > 0
```
Comment (Zethson) [inline main.nf]: Is there a neater way of testing whether or not a vcf  file was supplied? I'd rather not use if else logic, but a 'when:' statement.

**Comment by Zethson** [comment]
@Leon-Bichmann 

I guess it's ready to review/get tested by you now.
Predicted peptides, resolved neoepitopes and predicted peptides output now features a geneID column just after the peptide sequence.

**Comment by Leon-Bichmann** [comment]
Thanks, ill give it a trial run as soon as possible.

File: main.nf (hunk 9 above)
Comment (Leon-Bichmann) [reply main.nf]: For me this step is failing, I would go for when params.include_proteins_from_vcf

**Pull request merged**